Feature3: 구글 OAuth2.0 로그인 기능 완성

nAhYUN22 opened the pull request:
**개요**
Spring Security를 활용한 Google OAuth2.0 로그인 기능을 처음부터 구현하고, 세션 기반 인증 시스템을 구축했습니다.

**주요내용**
OAuth2 의존성 추가: Spring Security OAuth2 Client 라이브러리 통합
OAuth2 환경 설정: application-local.yml에 Google OAuth2 클라이언트 정보 및 환경 변수 구성
보안 설정 구현: 세션 기반 인증 및 CORS 정책을 포함한 SecurityConfig 작성
인증 핸들러: OAuth2 로그인 성공/실패 시 처리 로직 구현
사용자 정보 API: AuthController를 통한 인증된 사용자 정보 조회 엔드포인트 제공
DTO 설계: OAuth2 사용자 데이터 전달을 위한 UserInfo DTO 생성
문서화: OAuth 통합 과정 및 사용 방법에 대한 상세한 문서 작성
저장소 정리: .env.dev 및 .env.prod 파일을 저장소에서 제거하여 보안 강화
브랜치 통합: develop 브랜치를 feature/3에 머지하여 최신 코드 동기화

**기대효과**
Google 계정을 통한 간편한 소셜 로그인 기능 제공으로 사용자 경험 개선
Spring Security 기반의 안전한 인증 체계 구축
세션 관리를 통한 안정적인 사용자 상태 유지
CORS 설정으로 프론트엔드와의 안전한 통신 보장
환경 변수를 통한 설정 관리로 보안성 및 유연성 향상
상세한 문서화로 팀원들의 OAuth 기능 이해 및 유지보수 용이성 확보

nAhYUN22 committed:
feat: Implement Google OAuth2.0 login with Spring Security

- Add Spring Security OAuth2 Client dependencies
- Configure OAuth2 settings in application-local.yml with environment variables
- Create SecurityConfig with session-based authentication and CORS settings
- Implement OAuth2 authentication success/failure handlers
- Add AuthController for user info API endpoint
- Create UserInfo DTO for OAuth2 user data
- Add comprehensive OAuth integration documentation

nAhYUN22 committed:
chore: Remove .env.dev and .env.prod from repository

These files contain sensitive information and should not be tracked by Git.
They are already excluded by .gitignore (*.env pattern).

nAhYUN22 committed:
Merge branch 'develop' into feature/3

Long9725 commented:
**Claude finished @nAhYUN22's task** —— [View job](https://github.com/YYOAJUNG/budgetops-backend/actions/runs/19128300079)

---
### PR Review Complete ✅

#### Tasks
- [x] Read CLAUDE.md and review guidelines
- [x] Analyze PR changes
- [x] Review security configuration
- [x] Review OAuth handlers
- [x] Review controller and DTOs
- [x] Provide comprehensive feedback

---

## 리뷰 결과

이 PR은 Google OAuth2.0 로그인 기능을 Spring Security를 사용하여 구현한 것입니다. 전반적으로 깔끔한 구현이지만, 보안과 아키텍처 관점에서 몇 가지 개선이 필요합니다.

### 🔴 Critical Issues (보안 및 필수 수정 사항)

#### 1. CSRF 비활성화 보안 위험
**위치**: `SecurityConfig.java:30`
```java
.csrf(AbstractHttpConfigurer::disable)
```
**문제점**: 
- CSRF 보호가 완전히 비활성화되어 있습니다
- 세션 기반 인증(`SessionCreationPolicy.IF_REQUIRED`)을 사용하면서 CSRF를 끄면 CSRF 공격에 취약합니다
- OAuth2 로그인은 세션 기반이므로 CSRF 보호가 필수입니다

**권장사항**: 
- CSRF 보호를 활성화하거나, 최소한 특정 엔드포인트만 제외하는 방식으로 변경
- 또는 JWT 기반 stateless 인증으로 전환 후 CSRF 비활성화 고려

#### 2. CORS 설정 하드코딩
**위치**: `SecurityConfig.java:57`
```java
configuration.setAllowedOrigins(List.of("http://localhost:3000"));
```
**문제점**:
- 프론트엔드 URL이 하드코딩되어 있어 환경별 설정이 불가능합니다
- 프로덕션/스테이징/개발 환경에서 다른 URL을 사용할 수 없습니다

**권장사항**:
```java
@Value("${app.cors.allowed-origins}")
private List<String> allowedOrigins;
```
환경변수로 관리하여 배포 환경별로 다르게 설정

#### 3. 예외 정보 노출 (보안 취약점)
**위치**: `OAuth2AuthenticationFailureHandler.java:27`
```java
.queryParam("error", exception.getLocalizedMessage())
```
**문제점**:
- 예외의 상세 메시지가 쿼리 파라미터로 프론트엔드에 노출됩니다
- 내부 시스템 정보나 스택 트레이스가 포함될 수 있어 보안 위험이 있습니다
- 로그에는 상세 정보를 남기되, 사용자에게는 일반적인 메시지만 전달해야 합니다

**권장사항**:
```java
.queryParam("error", "authentication_failed")
// 또는 에러 코드 매핑하여 전달
```

#### 4. 로그에 민감 정보 출력
**위치**: `OAuth2AuthenticationSuccessHandler.java:29`
```java
log.info("OAuth2 로그인 성공 - email: {}, name: {}", email, name);
```
**문제점**:
- 이메일 주소는 개인정보이므로 프로덕션 로그에 출력하면 안 됩니다
- GDPR, 개인정보보호법 위반 가능성이 있습니다

**권장사항**:
```java
log.info("OAuth2 로그인 성공 - userId: {}", userId); // 이메일 대신 ID만
// 또는 debug 레벨로 변경
```

### 🟡 Important Issues (아키텍처 및 설계 개선)

#### 5. 비즈니스 로직 누락
**문제점**:
- OAuth2 인증 성공 후 사용자 정보를 DB에 저장하거나 업데이트하는 로직이 없습니다
- 실제 서비스에서는 인증된 사용자를 시스템에 등록해야 합니다
- 현재는 세션에만 정보가 저장되고, 영속화되지 않습니다

**권장사항**:
- `UserService` 생성하여 사용자 정보 저장/조회 로직 구현
- `OAuth2AuthenticationSuccessHandler`에서 서비스 호출
- 신규 사용자는 회원가입, 기존 사용자는 정보 업데이트

#### 6. 계층 구조 문제
**위치**: `AuthController.java:23-27`
```java
UserInfo userInfo = UserInfo.builder()
        .email(oAuth2User.getAttribute("email"))
        .name(oAuth2User.getAttribute("name"))
        .picture(oAuth2User.getAttribute("picture"))
        .build();
```
**문제점**:
- Controller에 비즈니스 로직(DTO 변환)이 있습니다
- `.claude/review/ARCHITECTURE.md`의 지침에 따르면 Controller는 요청/응답 변환만 담당해야 합니다

**권장사항**:
- Service 레이어를 만들어 사용자 정보 조회 로직 이동
- Controller → Service → Repository 구조 유지

#### 7. application.yml 중복 설정
**위치**: `application.yml`
- 동일한 OAuth2 설정이 `application-local.yml`과 `application.yml`에 중복되어 있습니다
- `server.port: 8080`이 두 번 선언되어 있습니다 (line 67, 89)

**권장사항**:
- 공통 설정은 `application.yml`에, 환경별 설정은 `application-{profile}.yml`에만

#### 8. Redirect URI 관리 문제
**위치**: Handler 클래스들
- `@Value("${app.oauth2.redirect-uri:http://localhost:3000/oauth/callback}")`
- 기본값이 하드코딩되어 있고, 각 Handler마다 동일한 설정을 주입받습니다

**권장사항**:
- 설정을 `@ConfigurationProperties`로 묶어서 관리
- 또는 공통 OAuth2 설정 클래스를 만들어 주입

### 🟢 Good Practices

1. ✅ OAuth2 client-id, client-secret을 환경변수로 관리하여 하드코딩 방지
2. ✅ Handler 클래스를 분리하여 성공/실패 케이스 명확하게 처리
3. ✅ Lombok 활용으로 보일러플레이트 코드 감소
4. ✅ Spring Security의 표준 방식을 따라 구현
5. ✅ 의존성 주입을 생성자 주입(`@RequiredArgsConstructor`)으로 처리

### 📝 Minor Issues

#### 9. 테스트 코드 부재
- OAuth 인증 flow에 대한 테스트가 없습니다
- 최소한 단위 테스트(Controller, Handler)는 작성하는 것을 권장합니다

**권장 테스트**:
```java
@WebMvcTest(AuthController.class)
@Import(SecurityConfig.class)
class AuthControllerTest {
    @Test
    @WithMockUser
    void getCurrentUser_인증된사용자_반환() { ... }
}
```

#### 10. API 문서화 부족
- `AuthController`에 Swagger 애노테이션이 없습니다
- `/api/auth/user` 엔드포인트에 대한 문서가 필요합니다

**권장**:
```java
@Operation(summary = "현재 인증된 사용자 정보 조회")
@ApiResponses(...)
@GetMapping("/user")
```

#### 11. 패키지 구조 개선 가능
현재: `oauth.controller`, `oauth.dto`, `oauth.handler`

**권장**: 
```
auth/
├── controller/
├── service/
├── domain/
├── dto/
└── config/
    ├── handler/
    └── SecurityConfig
```

도메인 중심으로 패키지를 구성하면 응집도가 높아집니다.

#### 12. UserInfo DTO 검증 부재
**위치**: `UserInfo.java`
- 필드에 대한 validation 애노테이션이 없습니다
- null 체크가 없어 NPE 발생 가능성이 있습니다

**권장**:
```java
@NotBlank(message = "이메일은 필수입니다")
private String email;
```

---

## 요약

### 반드시 수정해야 할 사항 (P0)
1. ⚠️ CSRF 보호 활성화 또는 stateless 인증으로 전환
2. ⚠️ 예외 메시지 노출 제거 (`OAuth2AuthenticationFailureHandler`)
3. ⚠️ 로그에서 이메일 등 민감 정보 제거
4. ⚠️ CORS 설정 환경변수화

### 강력히 권장하는 사항 (P1)
5. 사용자 정보 영속화 로직 추가 (UserService + Repository)
6. Controller의 비즈니스 로직을 Service로 이동
7. application.yml 중복 설정 정리
8. 테스트 코드 추가

### 개선 권장 사항 (P2)
9. 패키지 구조 개선
10. API 문서화 추가
11. DTO 검증 추가
12. 설정 관리 개선 (@ConfigurationProperties)

---

## 참고한 리뷰 가이드
- `.claude/review/README.md` - 리뷰 우선순위
- `.claude/review/SECURITY.md` - 보안 체크리스트
- `.claude/review/ARCHITECTURE.md` - 계층 구조 검증
- `.claude/review/CODING_STANDARDS.md` - 코딩 스타일

---

nAhYUN22 committed:
ci: Update deploy workflow to trigger on feature/3 branch

- Change deploy trigger from feature/1 to feature/3
- Add develop branch to pull request triggers

nAhYUN22 committed:
Merge branch 'feature/3' of github.com:YYOAJUNG/budgetops-backend into feature/3

nAhYUN22 committed:
refactor: Remove duplicate server.port configuration from application.yml

nAhYUN22 committed:
fix: application.yml 수정

nAhYUN22 committed:
fix:application-docker.yml 수정

nAhYUN22 committed:
fix: https:// to http://

nAhYUN22 committed:
security: Apply security improvements for OAuth2 authentication

- Enable CSRF protection with cookie-based token repository
- Ignore CSRF for OAuth2 endpoints to allow proper authentication flow
- Remove sensitive exception messages from failure handler
- Remove email/name logging from success handler to protect user privacy
- Configure CORS allowed origins via environment variable
- Use generic error message 'authentication_failed' instead of detailed exception
